### PR TITLE
Closes #2042 fix compute_dtf bug

### LINKED ISSUE #2042: Bug: `compute_dtf()` does not work correctly
### What happened?

If parts of the date in between are missing, e.g., day is missing but hour is present, `compute_dtf()` does not work correctly.

### Session Information

_No response_

### Reproducible Example

```
compute_dtf(dtc = "2022-06--T00:00", dt = ymd("2022-06-01"))
```
produces
```
[1] NA
```

### PR DESCRIPTION
**Pull request by zdz2101**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiraldev/devel/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/devel/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/cran-release/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Review by bms63 — COMMENTED** [review]
Are there any other variations that we need to be worried about here?

**Comment by zdz2101** [comment]
> Are there any other variations that we need to be worried about here?

For the most part, any non-standard-datetime formatting is caught by the `warn_if_invalid_dtc()` higher up in the code

**Comment by bms63** [comment]
> > Are there any other variations that we need to be worried about here?
> 
> For the most part, any non-standard-datetime formatting is caught by the `warn_if_invalid_dtc()` higher up in the code

But this one is not caught?

**Comment by zdz2101** [comment]
> > > Are there any other variations that we need to be worried about here?
> > 
> > 
> > For the most part, any non-standard-datetime formatting is caught by the `warn_if_invalid_dtc()` higher up in the code
> 
> But this one is not caught?

Something like 
`compute_dtf("2019---06T00:00", as.Date("2019-06-06"))` 
would've slipped through the cracks too outputting `NA` instead of the desired `"M"` 
but I added a check for the date-portion of a date-time string to properly determine if its the M or the D that needs to be imputed

File: R/derive_date_vars.R [hunk 1]
```diff
@@ -1149,12 +1150,23 @@ compute_dtf <- function(dtc, dt) {
...
 
+  # Find date portion
+  date_portion <- ifelse(grepl("T", dtc),
+    gsub("T", "", substr(dtc, 1, gregexpr("T", dtc))),
+    substr(dtc, 1, 10)
+  )
+  n_chr_date_portion <- nchar(date_portion)
+
+  # Location of the first instance of the double hyphen to determine if its month/day imputation
+  location_of_double_hyphen <- unlist(gregexpr("--", date_portion))
```
Comment (bundfussr) [inline R/derive_date_vars.R:1161]: This does not work for
```
> compute_dtf("2020----T00:00", ymd_hms("2020-05-15T00:00:00"))
[1] "M" NA 
```
Maybe we should use a different approach. Something like:
```
pdate <- get_partialdatetime(dtc)
impflags <- c(year = "Y", month = "M", day = "D")
impflags[[min(which(is.na(pdate)))]]
```

Comment (bundfussr) [reply R/derive_date_vars.R]: Unfortunately, there are still cases which are not covered:
```
> compute_dtf("-----T00:00", ymd_hms("2020-05-15T00:00:00"))
[1] NA
> compute_dtf("--05--T00:00", ymd_hms("2020-05-15T00:00:00"))
[1] NA
```
I don't mind which way we go if it works for all cases and is readable.

Comment (zdz2101) [reply R/derive_date_vars.R]: Are those acceptable formats? I'm pretty sure those are supposed to be caught by the 
`admiraldev:: is_valid_dtc()` and `admiraldev::warn_if_invalid_dtc()` calls. If that's the case, this becomes an `admiraldev` fix too

Comment (zdz2101) [reply R/derive_date_vars.R]: It looks like `warn_if_invalid_dtc()` wants to parse out any of the missing year stuff, but `is_valid_dtc()` basically allows it? 
image
image

Comment (bundfussr) [reply R/derive_date_vars.R]: In the beginning we ignored dates with missing elements in between (except dates like "2003---15"). The ideas was to release a first version quickly and handle the rare edge cases like "--12-15" later.
Cases like "--12-15" are explicitly mentioned in the ADaM IG. Thus I think we should cover them.

Comment (zdz2101) [reply R/derive_date_vars.R]: I believe I've added this portion, at least with your reprex's above. I shall make an issue in `admiraldev` to reflect the removal of those cases on the "NOT handled ...." messaging

**Pull request merged**